tuplex/codegen/src/FunctionRegistry.cc

```diff
@@ -1628,6 +1631,99 @@ namespace tuplex {
             return SerializableValue();
         }
 
+        SerializableValue FunctionRegistry::createIteratorRelatedSymbolCall(tuplex::codegen::LambdaFunctionBuilder &lfb,
+                                                                            llvm::IRBuilder<> &builder,
+                                                                            const std::string &symbol,
+                                                                            const python::Type &argsType,
+                                                                            const python::Type &retType,
+                                                                            const std::vector<tuplex::codegen::SerializableValue> &args,
+                                                                            IteratorInfo *iteratorInfo) {
+            if(symbol == "iter") {
+                return createIterCall(lfb, builder, argsType, retType, args);
+            }
+
+            if(symbol == "zip") {
+                return createZipCall(lfb, builder, argsType, retType, args, iteratorInfo);
+            }
+
+            if(symbol == "enumerate") {
+                return createEnumerateCall(lfb, builder, argsType, retType, args, iteratorInfo);
+            }
+
+            if(symbol == "next") {
+                return createNextCall(lfb, builder, argsType, retType, args, iteratorInfo);
+            }
+
+            return SerializableValue(nullptr, nullptr);
+        }
+
+        SerializableValue FunctionRegistry::createIterCall(LambdaFunctionBuilder &lfb, llvm::IRBuilder<> &builder,
+                                                           const python::Type &argsType, const python::Type &retType,
+                                                           const std::vector<tuplex::codegen::SerializableValue> &args) {
+            if(argsType.parameters().size() != 1) {
+                Logger::instance().defaultLogger().error("iter() currently takes single iterable as argument only");
+                return SerializableValue(nullptr, nullptr);
+            }
+
+            python::Type argType = argsType.parameters().front();
+            if(argType.isIteratorType()) {
+                // iter() call on a iterator. Simply return the iterator as it is.
+                return args.front();
+            }
+            return _icp->initIterContext(lfb, builder, argType, args.front());
```

Comment:
hm I see, we should prob. discuss whether there's a better way to refactor this and use some eloquent design pattern.

Reply:
postponed